### ISSUE TITLE
RFC: Jacobian of multiple return functions and structured Jacobian etc.?

### ISSUE THREAD

**Issue by Roger-luo** [FluxML/Zygote.jl#564]
I meet this problem while fixing @willtebbutt 's request in https://github.com/JuliaDiff/FiniteDifferences.jl/pull/54 since I hope we could have the same API in both packages and it seems this problem was not discussed before, I'll post what I'm thinking here

My previous implementation of the jacobian (as well as its original pytorch version) function won't work for multiple return value functions, and I actually never considered it. Now I start wondering which is better

## Option 1: non-structured

the most straightforward way is to convert all the output to a large vector. This is what's implemented in FiniteDifference, but it's not implemented in Zygote.

but this is definitely not something good in more serious calculation since this will cause a lot of intermediate allocations and it lost the original structure of the values

## Option 2: structured

It's more complicated but I feel it will benefit a lot of things

**Definition** *Adjoint type*: every type `T` has an adjoint type `∇(T)`, which is the type of its gradient, and an adjoint type `J(T, R)` which is its jacobian type, where `R` is its return type.

e.g for naive cases, it's usually the original type itself, the adjoint type of `Array` is `Array` and `Matrix`, the adjoint of `Numer` is `Numer` and `Matrix`, `Tuple` is `Tuple` and `Tuple` of `Matrix`. But for `struct`s, it's a `NamedTuple` and in my own case while using Flux, I actually had to define my own type of a model as `(index::Int, grad::Array)` to avoid type instability caused by auto generated gradient `Tuple{Nothing, Matrix}` and `Tuple{Matrix, Nothing}` so far this is just what Zygote is doing current.

first, as we all agree, the Jacobian of multivariant function `f(xs...)::Union{Array, Numer}` is a `Tuple`, where each element is the Jacobian matrix of each `x`, thus we could expand this definition a bit, the Jacobian of the multi-return function `f(x) -> ys...` is a tuple of each `y`'s jacobian, e.g

the jacobian of

```julia
f(x::Array)::Tuple{Array, Array}
```

is

```julia
Tuple{Matrix, Matrix}
```

now, a bit more general, the jacobian type of  a multi return value functions `f(xs...) -> ys...` where each `y` can be `AbstractArray` or `Numer` will have a double layered `Tuple` (also make sense if you take `Tuple` as some sort of array), e.g the jacobian of

```julia
f(x1, x2)::Tuple{Array, Array, Array}
```

is

```julia
Tuple{Tuple{Matrix, Matrix, Matrix}, Tuple{Matrix, Matrix, Matrix}}
```

so in conclusion, the jacobian of a multivariant multi-return function is

```julia
f(x1::T1,  x2::T2, xs...)::Tuple{R1, R2, ...} -> (y1, y2, ys...)
```

```julia
Tuple{
   Tuple{J(T1, R1), J(T1, R2), ...}, 
   Tuple{J(T2, R1), J(T2, R2), ...},
}
```

this idea also gives some other nice things, e.g for structures, it's jacobians are structured as well

```julia
struct Foo
    x::Float32
   y::Matrix{Float32}
end
```

will have gradient type of `NamedTuple{(:x, :y), Tuple{Float32, Matrix{Float32}}}` and jacobian type of `NamedTuple{(:x, :y), Tuple{Vector{Float32}, Matrix{Float32}}}`

**Comment by Roger-luo** [comment]
And one can always convert these structured Jacobians to one matrix by using a similar function like `to_vec` in `FiniteDifference` for both arguments and the Jacobian.
